### PR TITLE
Add gradient ascent optimization algorithms

### PR DESCRIPTION
**Pull request by salaast**
This is part of the effort to port the work by Krzysztof Choromanski, Mark Rowland, Vikas Sindhwani, Richard E. Turner, Adrian Weller:  "Structured Evolution with Compact Architectures for Scalable Policy Optimization", https://arxiv.org/abs/1804.02395

First step in Issue #248

### CODE REVIEW

File: compiler_opt/es/gradient_ascent_optimization_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+  def get_state(self):
+    return self.moving_average.tolist()
+
+  def set_state(self, state):
+    self.moving_average = np.asarray(state, dtype=np.float32)
+
+
+class AdamOptimizer(GAOptimizer):
+  """Class implementing ADAM gradient ascent optimizer."""
```
Comment (ebrevdo) [inline compiler_opt/es/gradient_ascent_optimization_algorithms.py:124]: Please describe the state vector here and in other optimizers' docstrings.  Looks like here it's [mmv1] + [mmv2] + [1] where the last dim corresponds to the current step?

**Comment by salaast** [comment]
> This is a piece of the total picture, right? Do you have a doc describing the entire GA plan?

The plan has been detailed in Issue #248

File: compiler_opt/es/gradient_ascent_optimization_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+Library of stateful gradient descent algorithms taking as input the gradient and
+current parameters, and output the new parameters.
+"""
+
+import abc
+import numpy as np
+
+
+# TODO(kchoro): Borrow JAXs optimizer library here. Integrated into Blackbox-v2.
+class GAOptimizer(metaclass=abc.ABCMeta):
```
[the remark this answers is not in the document]
Comment (salaast) [reply compiler_opt/es/gradient_ascent_optimization_algorithms.py:35]: The rename will be addressed after porting and is kept track of with Issue #254.
The signatures will be addressed in a type annotations PR after the initial port.

File: compiler_opt/es/gradient_ascent_optimization_algorithms_test.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+class GradientAscentOptimizationAlgorithmsTest(parameterized.TestCase):
+
+  @parameterized.parameters((np.asarray([1., 2., 3.], dtype=np.float32),),
+                            (np.asarray([1.1, 2.2, 3.3], dtype=np.float32),))
+  def test_momentum_set_state(self, state):
+    optimizer = gradient_ascent_optimization_algorithms.MomentumOptimizer(
+        0.1, 0.9)
+    optimizer.set_state(state)
+    recovered_state = optimizer.get_state()
+    np.testing.assert_array_equal(state, recovered_state)
```
[the remark this answers is not in the document]
Comment (ebrevdo) [reply compiler_opt/es/gradient_ascent_optimization_algorithms_test.py:47]: Please remember to fix this in the next PR.

**Review by ebrevdo — APPROVED** [review]

**Review by mtrofin — APPROVED** [review]

File: compiler_opt/es/gradient_ascent_optimization_algorithms.py (hunk 1 above)
Comment (salaast) [reply compiler_opt/es/gradient_ascent_optimization_algorithms.py]: Descriptions for the states have been added now

File: compiler_opt/es/gradient_ascent_optimization_algorithms_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (salaast) [reply compiler_opt/es/gradient_ascent_optimization_algorithms_test.py]: The instances have been changed to assert_array_almost_equal now

File: compiler_opt/es/gradient_ascent_optimization_algorithms.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+#
+# This is a port of the work by: Krzysztof Choromanski, Mark Rowland,
+# Vikas Sindhwani, Richard E. Turner, Adrian Weller:  "Structured Evolution
+# with Compact Architectures for Scalable Policy Optimization",
+# https://arxiv.org/abs/1804.02395
+#
+###############################################################################
+r"""Library of gradient ascent algorithms.
+
+Library of stateful gradient descent algorithms taking as input the gradient and
```
[the remark this answers is not in the document]
Comment (salaast) [reply compiler_opt/es/gradient_ascent_optimization_algorithms.py]: The text has been changed to gradient ascent in the most recent commit

**Pull request merged**